.github/workflows/main.yml

```diff
@@ -64,7 +92,71 @@ jobs:
     - run: cargo install --root ${{ runner.tool_cache }}/cargo-vet --version ${{ env.CARGO_VET_VERSION }} cargo-vet
     - run: cargo vet --locked
 
+    # common logic to cancel the entire run if this job fails
+    - run: gh run cancel ${{ github.run_id }}
+      if: failure() && github.event_name != 'pull_request'
+      env:
+        GH_TOKEN: ${{ github.token }}
+
+  # This job is a dependency of many of the jobs below. This calculates what's
+  # actually being run for this workflow. For example:
+  #
+  # * Pushes to branches, which is currently both pushes to merge queue branches
+  #   as well as release branches, perform full CI.
+  # * PRs to release branches (not `main`) run full CI.
+  # * PRs to `main` will only run a few smoke tests above plus some elements of
+  #   the test matrix. The test matrix here is determined dynamically by the
+  #   `./ci/build-test-matrix.js` script given the commits that happened and
+  #   the files modified.
+  determine:
+    name: Determine CI jobs to run
+    runs-on: ubuntu-latest
+    outputs:
+      run-full: ${{ steps.calculate.outputs.run-full }}
+      test-matrix: ${{ steps.calculate.outputs.test-matrix }}
+      test-capi: ${{ steps.calculate.outputs.test-capi }}
+      build-fuzz: ${{ steps.calculate.outputs.build-fuzz }}
+    steps:
+    - uses: actions/checkout@v3
+    - id: calculate
+      env:
+        GH_TOKEN: ${{ github.token }}
+      run: |
+        touch commits.log names.log
+        if [ "${{ github.event_name }}" != "pull_request" ]; then
+          run_full=true
```

Comment:
It might be worth adding a comment here that the intent is not that we re-run full CI for merges to `main`, but only for release branch and merge-queue builds.